stan/math/prim/meta/holder.hpp

```diff
@@ -0,0 +1,274 @@
+  }
+};
+
+}  // namespace math
+}  // namespace stan
+
+namespace Eigen {
+namespace internal {
+
+template <typename ArgType, typename... Ptrs>
+struct evaluator<stan::math::Holder<ArgType, Ptrs...>>
+    : evaluator_base<stan::math::Holder<ArgType, Ptrs...>> {
+  typedef stan::math::Holder<ArgType, Ptrs...> XprType;
+  typedef typename remove_all<typename nested_eval<ArgType, 1>::type>::type
+      ArgTypeNestedCleaned;
+  typedef typename XprType::CoeffReturnType CoeffReturnType;
+  typedef typename XprType::Scalar Scalar;
+  enum {
+    CoeffReadCost = evaluator<ArgTypeNestedCleaned>::CoeffReadCost,
+    Flags
+    = evaluator<ArgTypeNestedCleaned>::Flags
+      & (HereditaryBits | LinearAccessBit | PacketAccessBit | DirectAccessBit),
+    Alignment = evaluator<ArgTypeNestedCleaned>::Alignment,
+  };
+
+  evaluator<ArgTypeNestedCleaned> m_argImpl;
+
+  explicit evaluator(const XprType& xpr) : m_argImpl(xpr.m_arg) {}
+
+  EIGEN_STRONG_INLINE CoeffReturnType coeff(Index row, Index col) const {
+    return m_argImpl.coeff(row, col);
+  }
+  EIGEN_STRONG_INLINE CoeffReturnType coeff(Index index) const {
+    return m_argImpl.coeff(index);
+  }
+
+  EIGEN_STRONG_INLINE Scalar& coeffRef(Index row, Index col) {
+    return m_argImpl.coeffRef(row, col);
+  }
+  EIGEN_STRONG_INLINE Scalar& coeffRef(Index index) {
+    return m_argImpl.coeffRef(index);
+  }
+
+  template <int LoadMode, typename PacketType>
+  EIGEN_STRONG_INLINE PacketType packet(Index row, Index col) const {
+    return m_argImpl.template packet<LoadMode, PacketType>(row, col);
+  }
+  template <int LoadMode, typename PacketType>
+  EIGEN_STRONG_INLINE PacketType packet(Index index) const {
+    return m_argImpl.template packet<LoadMode, PacketType>(index);
+  }
+
+  template <int StoreMode, typename PacketType>
+  EIGEN_STRONG_INLINE void writePacket(Index row, Index col,
+                                       const PacketType& x) {
+    return m_argImpl.template writePacket(row, col, x);
+  }
+  template <int StoreMode, typename PacketType>
+  EIGEN_STRONG_INLINE void writePacket(Index index, const PacketType& x) {
+    return m_argImpl.template writePacket(index, x);
+  }
+};
+
+}  // namespace internal
+}  // namespace Eigen
+
+namespace stan {
+namespace math {
+
+/**
+ * Constructs a no-op operation that also holds pointer to some other
+ * expressions, allocated on heap. When the object is destructed those
+ * expressions will be deleted.
+ * @tparam T type of argument expression
+ * @tparam Ptrs types of pointers
+ * @param a argument expression
+ * @param ptrs pointers to objects the constructed object will own.
+ * @return holder operation
+ */
+template <typename T, typename... Ptrs,
+          std::enable_if_t<sizeof...(Ptrs) >= 1>* = nullptr>
+Holder<T, Ptrs...> holder(T&& arg, Ptrs*... pointers) {
+  return Holder<T, Ptrs...>(std::forward<T>(arg), pointers...);
+}
+// trivial case with no pointers constructs no holder object
+template <typename T>
+T holder(T&& arg) {
```

Comment:
Won't `T` as the return make a copy here?

Reply:
For lvalue input T will be lvalue ref, so no. For rvalue input yes, but only of the compiler does not do RVO.

Reply:
Yes good call!

https://godbolt.org/z/nzkf2m